libkbfs/file_data.go

```diff
@@ -97,6 +122,341 @@ func (fd *fileData) getFileBlockAtOffset(ctx context.Context,
+	}
+
+	type resp struct {
+		pathsFromRoot [][]parentBlockAndChildIndex
+		blocks        map[BlockPointer]*FileBlock
+		nextBlockOff  int64
+	}
+
+	// Search all of the in-range child blocks, and their child
+	// blocks, etc, in parallel.
+	respChans := make([]<-chan resp, 0, len(pblock.IPtrs))
+	eg, groupCtx := errgroup.WithContext(ctx)
+	nextBlockOff := int64(-1)
+	for i, iptr := range pblock.IPtrs {
+		// Some byte of this block is included in the left side of the
+		// range if `startOff` is less than the largest byte offset in
+		// the block.
+		inRangeLeft := true
+		if i < len(pblock.IPtrs)-1 {
+			inRangeLeft = startOff < pblock.IPtrs[i+1].Off
+		}
+		if !inRangeLeft {
+			continue
+		}
+		// Some byte of this block is included in the right side of
+		// the range if `endOff` is bigger than the smallest byte
+		// offset in the block (or if we're explicitly reading all the
+		// data to the end).
+		inRangeRight := endOff == -1 || endOff > iptr.Off
+		if !inRangeRight {
+			// This block is the first one past the offset range
+			// amount the children.
+			nextBlockOff = iptr.Off
+			break
+		}
+
+		ptr := iptr.BlockPointer
+		childIndex := i
+		respCh := make(chan resp, 1)
+		respChans = append(respChans, respCh)
+		eg.Go(func() error {
+			block, err := fd.goGetter(groupCtx, fd.kmd, ptr, fd.file)
+			if err != nil {
+				return err
+			}
+			// Recurse down to the level of the child.
+			pfr, blocks, nextBlockOff, err := fd.getLeafBlocksForOffsetRange(
+				groupCtx, block, startOff, endOff, prefixOk)
+			if err != nil {
+				return err
+			}
+
+			// Append self to the front of every path and add `block` to
+			// the blocks map.
+			var r resp
+			for _, p := range pfr {
+				newPath := append([]parentBlockAndChildIndex{{
+					pblock:     pblock,
+					childIndex: childIndex,
+				}}, p...)
+				r.pathsFromRoot = append(r.pathsFromRoot, newPath)
+			}
+			r.blocks = make(map[BlockPointer]*FileBlock)
+			r.blocks[ptr] = block
+			for ptr, block := range blocks {
+				r.blocks[ptr] = block
+			}
+			r.nextBlockOff = nextBlockOff
+			respCh <- r
+			return nil
+		})
+	}
+
+	// If we are ok with just getting the prefix, don't treat a
+	// deadline exceeded error as fatal.
+	if err := eg.Wait(); err != nil && (!prefixOk ||
+		err != context.DeadlineExceeded) {
+		return nil, nil, 0, err
+	}
+
+	blocks = make(map[BlockPointer]*FileBlock)
+outer:
+	for i, respCh := range respChans {
+		select {
+		case r := <-respCh:
+			pathsFromRoot = append(pathsFromRoot, r.pathsFromRoot...)
+			for ptr, block := range r.blocks {
+				blocks[ptr] = block
+			}
+			if i == len(respChans)-1 {
+				// Figure out the offset of the next data block after
+				// the last in-range one.  It's the minimum of the
+				// next one at this level and the next one at the
+				// child's level, not counting -1 (which means we are
+				// returning the the last block at the given level).
+				if (r.nextBlockOff != -1 && r.nextBlockOff < nextBlockOff) ||
```

Comment:
I was confused by this for a while. With the comment re. renaming `nextBlockOff` to `nextBlockOffThisLevel` above, might be clearer to have a `nextBlockOffChildLevel` variable and set it here, and then have

```
// Figure out the offset ...
nextBlockOffset = (min calculation using nextBlockOffThisLevel and nextBlockOffChildLevel)
return ..., nextBlockOffset, ...
```

outside the loop